pkg/operator/invalidcertscontroller/invalid_certs_controller.go

```diff
@@ -0,0 +1,145 @@
+
+	prometheusapi "github.com/prometheus/client_golang/api"
+	prometheusv1 "github.com/prometheus/client_golang/api/prometheus/v1"
+)
+
+type invalidCertsController struct {
+	operatorClient          v1helpers.OperatorClient
+	prometheusQuery         string
+	newPrometheusClientFunc func() (prometheusv1.API, idleConnectionCloser, error)
+}
+
+// NewInvalidCertsController creates a new invalid certificates controller using an in-cluster prometheus client.
+// The given service CA path will be used to read out the CA bundle being trusted by the prometheus client.
+//
+// The controller executes the given prometheus query every minute. The resulting value is interpreted as the amount
+// of invalid certificates. If the value is >=1 the `InvalidCertsDetected` condition state will be set.
+//
+// The supported prometheus query result types are:
+// - vector: only the first sample of the vector will be used for evaluation.
+// - scalar: the returned value will be used.
+func NewInvalidCertsController(operatorClient v1helpers.OperatorClient, recorder events.Recorder, serviceCAPath string, prometheusQuery string) factory.Controller {
+	c := &invalidCertsController{
+		operatorClient:  operatorClient,
+		prometheusQuery: prometheusQuery,
+		newPrometheusClientFunc: func() (prometheusv1.API, idleConnectionCloser, error) {
+			return newInClusterPrometheusClient(serviceCAPath)
+		},
+	}
+
+	return factory.New().
+		WithSync(c.sync).
+		ResyncEvery(time.Minute).
+		ToController("InvalidCertsController", recorder.WithComponentSuffix("invalid-certs-controller"))
+}
+
+func (c *invalidCertsController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	client, closer, err := c.newPrometheusClientFunc()
+	if err != nil {
+		return fmt.Errorf("error instantiating prometheus client: %w", err)
+	}
+	// if idle connections won't be closed, memory leaks will occur.
+	defer closer.CloseIdleConnections()
+
+	result, _, err := client.Query(ctx, c.prometheusQuery, time.Now())
+	if err != nil {
+		return fmt.Errorf("error executing prometheus query: %w", err)
+	}
+
+	var value model.SampleValue
+	switch result.Type() {
+	case model.ValVector:
+		value = result.(model.Vector)[0].Value
+	case model.ValScalar:
+		value = result.(*model.Scalar).Value
+	default:
+		return fmt.Errorf("unexpected prometheus query return type: %T", result)
+	}
+
+	_, _, err = v1helpers.UpdateStatus(ctx, c.operatorClient, v1helpers.UpdateConditionFn(newInvalidCertsCondition(float64(value))))
+	return err
+}
+
+func newInvalidCertsCondition(invalidCertsCount float64) operatorv1.OperatorCondition {
+	condition := operatorv1.OperatorCondition{
+		Type:   "InvalidCertsUpgradeable",
+		Status: operatorv1.ConditionTrue,
+	}
+
+	if invalidCertsCount >= 1.0 {
+		condition.Status = operatorv1.ConditionFalse
+		condition.Reason = "InvalidCertsDetected"
+		condition.Message = fmt.Sprintf("%d invalid certificates were detected", int(invalidCertsCount))
```

Comment:
this needs to be self-explanatory - what's an invalid cert?

Reply:
the underlying query might have different semantics on what "invalid" means. suggestion: how about adding some context string in the constructor?

Reply:
for our use case, "invalid" means "common name is used vs. SAN fields".

Reply:
If you wanted to have the controller more generic, you'd have to allow for specifying custom checks that each generate their own condition.

The controller as of now is very specific to a single problem so you don't need to worry about adding context strings in the constructor and may just as well hard-code it here.

Reply:
you have a good point, let's make this controller then very specific to the "CN field used vs. SAN field" issue. I wonder if the underlying query should be more opinioated then, @dgrisonnet do you maybe have opinions?

Reply:
I personally really like the idea of keeping it generic because if we want to take into account other errors in the future, we will have to write a new controller which will most likely have a very similar logic. With that in mind, what do you think about:
- Decouple the Prometheus query part and create a PrometheusQuery controller whose sole role would be to execute queries and return values
- Use PrometheusQuery controllers in the invalidCertsController to get specific information about invalid certs and update status depending on the query and the values returned by the PrometheusQuery controllers.

That would allow extending the invalidCertsController with new checks whenever we want.

Reply:
it seems we converged against consensus, now reflected in code.